api/configuration_services.proto

```diff
@@ -57,6 +58,89 @@ message GetPhyVersionsResponse {
   repeated VersionInfo version_info = 2;
 }
 
+message ListBandsRequest {
+  // Optional Band ID to filter the results.
+  // If unused, all supported Bands and their versions are returned.
+  string band_id = 1;
+}
+
+message BandDescription {
+  string id = 1;
+
+  message Beacon {
+    DataRateIndex data_rate_index = 1;
+    string coding_rate = 2;
+    bool inverted_polarity = 3;
+    reserved 4; //ComputeFrequency
+  }
+
+  Beacon beacon = 2;
+  uint64 ping_slot_frequency = 3;
+  uint32 max_uplink_channels = 4;
+
+  message Channel {
+    uint64 frequency = 1;
+    DataRateIndex min_data_rate = 2;
+    DataRateIndex max_data_rate = 3;
+  }
+
+  repeated Channel uplink_channels = 5;
+  uint32 max_downlink_channels = 6;
+  repeated Channel downlink_channels = 7;
+
+  message SubBandParameters {
+    uint64 min_frequency = 1;
+    uint64 max_frequency = 2;
+    float duty_cycle = 3;
+    float max_eirp = 4;
+  }
+  
+  repeated SubBandParameters sub_bands = 8;
+
+  message DataRate {
+    DataRate rate = 1;
+    reserved 2; //MaxMACPayloadSize
+  }
+
+  map<uint32, DataRate> data_rates = 9;
+
+  uint64 freq_multiplier = 10;
+  bool implements_cf_list = 11;
+  CFListType cf_list_type = 12;
+  google.protobuf.Duration receive_delay_1 = 13;
+  google.protobuf.Duration receive_delay_2 = 14;
+  google.protobuf.Duration join_accept_delay_1 = 15;
+  google.protobuf.Duration join_accept_delay_2 = 16;
+  uint64 max_fcnt_gap = 17;
+  bool supports_dynamic_adr = 18;
+  ADRAckLimitExponent adr_ack_limit = 19;
+  bool tx_param_setup_req_support = 20;
+  float default_max_eirp = 21;
+  string lora_coding_rate = 22;
+  reserved 23; //Rx1Channel
+  reserved 24; //Rx2DataRate
+  reserved 25; //GenerateChMasks
+  reserved 26; //ParseChMask
+
+  message Rx2Parameters {
+    DataRateIndex data_rate_index = 1; 
+    uint64 frequency = 2;
+  }
+
+  Rx2Parameters default_rx2_parameters = 27;
+
+  message DwellTime {
+    bool uplinks = 1;
+    bool downlinks = 2;
+  }
+
+  DwellTime boot_dwell_time = 28;
+}
+
+message ListBandsResponse {
+  repeated BandDescription bands = 1;
```

Comment:
While the `ListFrequencyPlansResponse` indeed returns a repeated result, here we have two levels of indirection: first the band ID, then the `PHYVersion`.

Since in protobuf we cannot have map of map, let's wrap `map<PHYVersion, BandDescription>` into a message and then make `map<string, that-wrapper>` part of the response.

Reply:
Changed the `Wrapper` to be of the kind `map<string, BandDescription>` as there is no support for an `enum` as key in maps.